### ISSUE TITLE
feat(learning_rate): Add NoamLR — Transformer warm-up schedule (Vaswani 2017)

### ISSUE THREAD

**Issue by sanskarpan** [sanskarpan/Optimization#35]
## Description
The learning rate schedule from 'Attention Is All You Need'.

```
lr(t) = scale · d_model^(-0.5) · min(t^(-0.5), t · warmup_steps^(-1.5))
```
Increases linearly for t < warmup_steps, then decreases ∝ 1/√t. Peaks at t = warmup_steps.

## Acceptance Criteria
- [ ] Monotone increasing before `warmup_steps`
- [ ] Peak at exactly `warmup_steps`
- [ ] Monotone decreasing after `warmup_steps`
- [ ] `d_model <= 0` or `warmup_steps <= 0` raise `ValueError`
- [ ] Step 0 handled (no division by zero)
